File: src/des-pipe/likelihood/planck/clik/clik_gibbs.c
#include "clik.h"
#include "clik_helper.h"
#include <errno.h>
#include <string.h>

typedef struct {
  int handle;
  int handle_transition;
  int ltrans,lmin;
  } gibbs;


void free_gibbs(void **none) {
  gibbs *gb;

  
  gb = *none;
  gibbs_extra_free_(&(gb->handle));
  if (gb->handle_transition!=-1) {
    gibbs_extra_free_(&(gb->handle_transition));
  }
}

double gibbs_lkl(void* none, double* pars, error **err) {
  double lkl;
  gibbs *gb;

  gb = none;
  gibbs_extra_lkl_(&lkl,&gb->handle,pars);
  
  if (gb->handle_transition!=-1) {
    double lkl_trans;
    gibbs_extra_lkl_(&lkl_trans,&gb->handle_transition,&(pars[gb->ltrans-gb->lmin]));
    
    lkl -= lkl_trans;
  }

  return lkl;
}

cmblkl* clik_gibbs_init(cldf *df, int nell, int* ell, int* has_cl, double unit,double* wl, double *bins, int nbins, error **err) {
  char directory_name[4096],pwd[4096],pwd2[4096];
  int status;
  int bok;
  cmblkl *cing;
  int mlmax;
  char dir_data[2048];
  int ldd;
  int lmin,lmax;
  int firstchain,lastchain,firstsample,lastsample,step;
  int hk;
  gibbs *gb;

  lmin = ell[0];
  lmax = ell[nell-1];
  
  // get data and change dir
  cldf_external(df,directory_name,pwd,err);
  forwardError(*err,__LINE__,NULL);
  
  firstchain = cldf_readint(df,"firstchain",err);
  forwardError(*err,__LINE__,NULL);
  lastchain = cldf_readint(df,"lastchain",err);
  forwardError(*err,__LINE__,NULL);
  firstsample = cldf_readint(df,"firstsample",err);
  forwardError(*err,__LINE__,NULL);
  lastsample = cldf_readint(df,"lastsample",err);
  forwardError(*err,__LINE__,NULL);
  step = cldf_readint(df,"step",err);
  forwardError(*err,__LINE__,NULL);


  memset(dir_data,' ',sizeof(char)*2048);
  sprintf(dir_data,"data/");
  dir_data[5] = ' ';
  ldd = 5;
  
  gb = malloc_err(sizeof(gibbs),err);
  forwardError(*err,__LINE__,NULL);
  
  gb->handle = 0;
  gb->handle_transition = -1;


  //call
  gibbs_extra_parameter_init_(&(gb->handle),dir_data,&ldd,&lmin,&lmax,&firstchain,&lastchain,&firstsample,&lastsample,&step);
  testErrorRetVA(gb->handle<=0,-43255432,"handle return is negative (got %d)",*err,__LINE__,NULL,gb->handle);

  hk = cldf_haskey(df,"ltrans",err);
  forwardError(*err,__LINE__,NULL);
  if (hk == 1) {
    int ltrans;
  
    gb->handle_transition = 1;
    ltrans = cldf_readint(df,"ltrans",err);
    forwardError(*err,__LINE__,NULL);
    gb->lmin = lmin;
    gb->ltrans = ltrans;
    
    gibbs_extra_parameter_init_(&(gb->handle_transition),dir_data,&ldd,&ltrans,&lmax,&firstchain,&lastchain,&firstsample,&lastsample,&step);
    testErrorRetVA(gb->handle<=0,-43255432,"handle return is negative (got %d)",*err,__LINE__,NULL,gb->handle);

  }

  cldf_external_cleanup(directory_name,pwd,err);
  forwardError(*err,__LINE__,NULL);
  
  cing = init_cmblkl(gb, &gibbs_lkl, 
                     &free_gibbs,
                     nell,ell,
                     has_cl,ell[nell-1],unit,wl,0,bins,nbins,0,err);
  forwardError(*err,__LINE__,NULL);

  return cing;
}

File: src/des-pipe/likelihood/planck/clik/clik_gibbs.f90
MODULE GIBBS_EXTRA

	IMPLICIT NONE

	INTEGER,dimension(100):: CLIK_LMAX,CLIK_LMIN
	real(8),dimension(2:1000) :: cltt
	real(8),parameter:: PI    = 3.141592653589793238462643383279502884197

END MODULE GIBBS_EXTRA




SUBROUTINE GIBBS_EXTRA_FREE(handle)
	use comm_br_mod
	use GIBBS_EXTRA
	INTEGER,intent(in)::handle
	
	call comm_br_deallocate_object(handle)
	
	clik_lmin(handle) = -1
	clik_lmax(handle) = -1
	
END SUBROUTINE 	GIBBS_EXTRA_FREE

SUBROUTINE GIBBS_EXTRA_LKL(LKL,handle,CL)
	use comm_br_mod
	use GIBBS_EXTRA
	
	REAL(8),INTENT(OUT)::LKL
	INTEGER,intent(in)::handle
	REAL(8),INTENT(IN),DIMENSION(0:CLIK_LMAX(handle)-CLIK_LMIN(handle))::CL
	INTEGER::i,cur

	!TT
	cur = 0
	cltt = 0
	DO i = clik_lmin(handle),clik_lmax(handle)
		cltt(i)=CL(cur)*(i*(i+1.))/2./PI
		cur = cur + 1
	END DO	

	LKL = comm_br_compute_lnL(cltt(2:clik_lmax(handle)),handle)		
	
END SUBROUTINE 	GIBBS_EXTRA_LKL



SUBROUTINE GIBBS_EXTRA_PARAMETER_INIT(handle,datadir,l_datadir,lmin,lmax,firstchain,lastchain,firstsample,lastsample,step)
	use comm_br_mod
	use GIBBS_EXTRA
	
	INTEGER,INTENT(IN)::l_datadir,lmin,lmax,firstchain,lastchain,firstsample,lastsample,step
	character(len=l_datadir)::datadir
	INTEGER,INTENT(OUT)::handle
	character(len=1000)::sigma_file,cl_file,data_dir

		
	data_dir = TRIM(datadir)
	sigma_file = TRIM(data_dir)//"/sigma.fits"
	cl_file = TRIM(data_dir)//"/cl.dat"

	handle = 0

	call comm_br_initialize_object(sigma_file, cl_file, lmin, lmax, firstchain, lastchain, &
       & firstsample, lastsample, step, handle)
	
	clik_lmin(handle) = lmin
	clik_lmax(handle) = lmax
		
END SUBROUTINE 	GIBBS_EXTRA_PARAMETER_INIT
